test/suite/test_checkpoint31.py

```diff
@@ -0,0 +1,120 @@
+
+# test_checkpoint31.py
+#
+# Test that skipping in-memory reconciled deleted pages as part of the tree walk.
+
+class test_checkpoint(wttest.WiredTigerTestCase):
+    conn_config = 'cache_size=500MB,statistics=(all)'
+
+    format_values = [
+    # FLCS doesn't support skipping pages based on aggregated time window.
+    #    ('column-fix', dict(key_format='r', value_format='8t',
+    #        extraconfig=',allocation_size=512,leaf_page_max=512')),
+        ('column', dict(key_format='r', value_format='S',
+            extraconfig=',allocation_size=512,leaf_page_max=512')),
+        ('string_row', dict(key_format='S', value_format='S',
+             extraconfig=',allocation_size=512,leaf_page_max=512')),
+    ]
+    
+    scenarios = make_scenarios(format_values)
+
+    def check(self, ds, nrows, value):
+        cursor = self.session.open_cursor(ds.uri)
+        count = 0
+        for k, v in cursor:
+            self.assertEqual(v, value)
+            count += 1
+        self.assertEqual(count, nrows)
+        cursor.close()
+
+    def test_checkpoint(self):
+        uri = 'table:checkpoint31'
+        nrows = 10000
+
+        # Create a table.
+        ds = SimpleDataSet(
+            self, uri, 0, key_format=self.key_format, value_format=self.value_format,
+            config=self.extraconfig)
+        ds.populate()
+
+        value_a = "aaaaa" * 100
+
+        # Write some initial data.
+        cursor = self.session.open_cursor(ds.uri, None, None)
+        for i in range(1, nrows + 1):
+            self.session.begin_transaction()
+            cursor[ds.key(i)] = value_a
+            self.session.commit_transaction()
+
+        # Create a reader transaction that will not be able to see what happens next.
+        # We don't need to do anything with this; it just needs to exist.
+        session2 = self.conn.open_session()
+        session2.begin_transaction()
+
+        # Now remove all data.
+        for i in range(100, nrows + 1):
+            self.session.begin_transaction()
+            cursor.set_key(ds.key(i))
+            self.assertEqual(cursor.remove(), 0)
+            self.session.commit_transaction()
+
+        # Checkpoint.
+        self.session.checkpoint()
+
+        # Get the existing cache eviction force delete statistic value.
+        stat_cursor = self.session.open_cursor('statistics:', None, None)
+        prev_cache_eviction_force_delete = stat_cursor[stat.conn.cache_eviction_force_delete][2]
+        stat_cursor.close()
+
+        # Now read the removed data.
+        self.session.breakpoint()
+        self.check(ds, 99, value_a)
+
+        # Get the new cache eviction force delete statistic value.
+        stat_cursor = self.session.open_cursor('statistics:', None, None)
+        cache_eviction_force_delete = stat_cursor[stat.conn.cache_eviction_force_delete][2]
+        stat_cursor.close()
+
+        self.assertGreater(cache_eviction_force_delete, prev_cache_eviction_force_delete)
+
+        # Tidy up.
```

Comment:
At this point, I am assuming that all the pages with deleted items (non-obsolete) should have been deleted the page should have evict soon flag and resetting the cursor should evict all these pages

Can we add a few more things to this test - 

1. Can we use the read using (session2) and ensure that skipped page count is 0?
2. Can we create a new session and ensure that skipped page could is greater than 0?

Reply:
skipped count cannot be zero, because we evict the page when all the updates on a page are removed. In this scenario, it is possible at least two pages with not completely removed, so we cannot assert that skipped count can be zero.

Yes, the pages can be greater than zero. Currently, we do not have stat to count the skipped pages, I will add the stat and update the test accordingly.

Reply:
Added a new statistic and validated it with the test.